### ISSUE TITLE
Use the library with custom job script or already submitted jobs

### ISSUE THREAD

**Issue by ekouts** [ExaWorks/psi-j-python#241]
Would it be possible to allow users to set directly the job script when creating the `JobSpec` object? I understand that in this way it would be tied to one scheduler but it would be nice to be able to use the rest of the library's features, like querying for jobs with callbacks and cancelling them through a python program.
Similarly would it be possible to use the library with a job that was submitted independently (possibly by providing the jobid of the scheduler) when creating the `Job` object?

**Comment by andre-merzky** [comment]
@ekouts: we do as of yet not support user provided job scripts.  That approach would also go against the abstractions psij tries to provide.  However, we'll discuss that request and ping back.

 >  Similarly would it be possible to use the library with a job that was submitted independently (possibly by providing the jobid of the scheduler) when creating the Job object?
 
 This is indeed supported:
 ```py
#!/usr/bin/env python3

import sys
import psij

native_id = sys.argv[1]
job = psij.Job()
jex = psij.JobExecutor.get_instance('slurm')
jex.attach(job, native_id)
status = job.wait()
```

**Comment by ekouts** [comment]
Great thanks for the quick answer! I tried the example and it worked indeed. I also set `PSIJ_BATCH_KEEP_FILES` to `True` in order to avoid an error message about failing to clean up the submit script. 
Let me know when you have any updates about the job script. In my case I would like to use the library as an abstraction for the submission / query / cancellation commands for different schedulers so it would be simpler if I could point to a script.

**Comment by hategan** [comment]
> we do as of yet not support user provided job scripts. That approach would also go against the abstractions psij tries to provide. However, we'll discuss that request and ping back.

This would indeed go against the idea of abstracting job submission. However, we do allow pass-throughs in various places (a notable example are the custom attributes) because we do recognize that some advance usage might require things beyond what the abstraction provides. So it seems to me like there is some precedent for breaking the abstraction to allow advanced usage. Furthermore, when providing a custom batch script, psi/j can still be useful in providing an API to monitor the resulting jobs.

The above is predicated on the idea that the abstraction we provide does not cover all the features that some scheduler has. While this is unavoidable, it may be good to ask if there is a particular feature that @ekouts needs that could be added to PSI/J so that, at least in this case, custom scripts would not be necessary.

**Comment by ekouts** [comment]
Thanks for taking the time to reply. I am not aware of any features that are missing from a scheduler. In our case we were considering using the library as a backend for a REST API: https://firecrest-api.cscs.ch
In firecrest the user is supposed to provide the batch script directly and only Slurm is supported at the moment. Even though we have a small number of scheduler commands that need to be executed it would make sense to take advantage of the work that is already done in psij, and possibly contribute if there is need in the future.
We understand though that currently PSI/J is meant to be used on a different level of abstraction and may not be the right solution for us in the end. We wanted to understand if there are plans to expand a bit the capabilities of the library to fit a more generic use case.

**Comment by andre-merzky** [comment]
I want to propose the following approach to address this issue:  the `psij.Job` constructor obtains an alternative `raw_spec` parameter instead of a `psij.JobSpec` which represents an executor implementation specific blob. For the `slurm` executor that blobcould be a Slurm batch script, for the `flux` executor that could be a Flux job spec (as json?), for the local executor an executable path.  No other parameters are allowed.  On `psij.JobExecutor.submit(job)`, an `ValueError` would be raised if that specific executor cannot interpret the blob and the job goes to `FAILED`.  Otherwise the job enters the normal state progression (`QUEUED` etc).

While this is a change to the API, it is the least intrusive change I can think of.  It has the disadvantage of allowing executor specific job instances which are invalid to other executors.  An alternative might be to add an `JobExecutor.submit_raw(job, blob)`, but I find that somewhat adverse to the design we have currently, and I don't like the possibility of additional failure modes (the job could have a spec already).

Feedback welcome (@hategan @kylechard @mtitov @jameshcorbett) - otherwise I can start on a PR.

**Comment by hategan** [comment]
So the batch scheduler executors (Slurm, PBS, etc.) initialize a SubmitScriptGenerator in their constructors, currently with some instance of TemplatedScriptGenerator. One could simply say `executor.generator = CustomScriptGenerator(my_path)`. While simple, the issue, as identified by Andre, is that this does not trickle up to the spec in any way, making it an implementation specific solution when we might want a more generic one.

The fundamental issue is that we cannot add a typesafe API parameter to the specification without disclosing implementation details, so we may have to chose somewhere between something like an untyped `raw_spec` that is interpreted by executor implementations in mysterious ways or prescribing details of how submit scripts are generated in the specification, hopefully in a minimal way.

@andre-merzky, perhaps let's not rush this; we can give it a little bit more thought before starting work or, perhaps better, adding it to the plans for a specific release.

**Comment by andre-merzky** [comment]
> The fundamental issue is that we cannot add a typesafe API parameter to the specification without disclosing implementation details, so we may have to chose somewhere between something like an untyped `raw_spec` that is interpreted by executor implementations in mysterious ways or prescribing details of how submit scripts are generated in the specification, hopefully in a minimal way.

Yeah, that's the crux, and I don't like that either.  It might be the smaller of two evils compared to uses using some 3rd party code or homegrown solution to submit programmatically and then using `attach(job, native_id)`.

> @andre-merzky, perhaps let's not rush this; we can give it a little bit more thought before starting work or, perhaps better, adding it to the plans for a specific release.

:+1:

**Comment by jameshcorbett** [comment]
I'm on board with not rushing it but I thought I would add my thoughts.

> An alternative might be to add an JobExecutor.submit_raw(job, blob)

I think I actually like this idea @andre-merzky . The different JobExecutors could have their own type signatures for the method, e.g. a `Path` to a submit script for Slurm/LSF/PBS/Cobalt and for Flux a jobspec string. And I think the `if job.spec is not None: raise Exception()` logic would actually be both simple to implement and to understand. By contrast making a `Job` have two mutually exclusive attributes (`spec` and `raw_spec`) seems confusing to me. 

My only concern would be that maybe there would be some `JobSpec` attributes we would want to capture even when submitting a user-generated batch script. Like imagine if some `sbatch` options couldn't be put in a script but had to be passed on the command line, a user might want to pass a custom script and ALSO tell PSI/J to add `--foo --bar` flags when it submitted the job.

**Comment by andre-merzky** [comment]
> My only concern would be that maybe there would be some `JobSpec` attributes we would want to capture even when submitting a user-generated batch script.

Yeah, that would be a concern indeed. That mix can become messy very quickly and I would suggest to not consider supporting the mixed approach unless a compelling use case arises.
